From 1db56c54b7a51bc8e7005a70f64ffe95f9998e12 Mon Sep 17 00:00:00 2001
From: Gaurav K Singh <gaurav.k.singh@intel.com>
Date: Thu, 11 Jun 2015 13:42:43 +0530
Subject: [PATCH] FOR_UPSTREAM [VPG]: drm/i915: Add error handling while
 creating GEM buffer

Error handling conditions while creating GEM buffer for DSI
command mode taken care now.

Change-Id: Ic8aa9751fff75bbe1545887b8b72c98700be26a6
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-7482
Signed-off-by: Gaurav K Singh <gaurav.k.singh@intel.com>
---
 drivers/gpu/drm/i915/intel_dsi.c | 17 ++++++++++++-----
 1 file changed, 12 insertions(+), 5 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_dsi.c b/drivers/gpu/drm/i915/intel_dsi.c
index e2a57d3..2ad10dd 100644
--- a/drivers/gpu/drm/i915/intel_dsi.c
+++ b/drivers/gpu/drm/i915/intel_dsi.c
@@ -371,6 +371,7 @@ static void intel_dsi_pre_enable(struct intel_encoder *encoder)
 	u32 temp;
 	u32 val;
 	u32 count = 1;
+	int ret;
 
 	DRM_DEBUG_KMS("\n");
 
@@ -381,13 +382,14 @@ static void intel_dsi_pre_enable(struct intel_encoder *encoder)
 			return;
 		}
 
-		i915_gem_object_set_cache_level(intel_dsi->gem_obj,
+		ret = i915_gem_object_set_cache_level(intel_dsi->gem_obj,
 							I915_CACHE_LLC);
+		if (ret)
+			goto err_unref;
 
-		if (i915_gem_obj_ggtt_pin(intel_dsi->gem_obj, 4096, 0)) {
-			DRM_ERROR("MIPI command buffer GTT pin failed");
-			return;
-		}
+		ret = i915_gem_obj_ggtt_pin(intel_dsi->gem_obj, 4096, 0);
+		if (ret)
+			goto err_unref;
 
 		intel_dsi->cmd_buff =
 				kmap(sg_page(intel_dsi->gem_obj->pages->sgl));
@@ -484,6 +486,11 @@ static void intel_dsi_pre_enable(struct intel_encoder *encoder)
 	/* Enable port in pre-enable phase itself because as per hw team
 	 * recommendation, port should be enabled befor plane & pipe */
 	intel_dsi_send_enable_cmds(encoder);
+	return;
+
+err_unref:
+	drm_gem_object_unreference(&intel_dsi->gem_obj->base);
+	return;
 }
 
 static void intel_dsi_enable(struct intel_encoder *encoder)
-- 
1.9.1

